File: matrix-multiplication/matrix_multiplication.py
def matrix_multiply(A, B):
    """Standard matrix multiplication O(n^3)"""
    rows_A, cols_A = len(A), len(A[0])
    rows_B, cols_B = len(B), len(B[0])
    
    if cols_A != rows_B:
        raise ValueError("Cannot multiply matrices: incompatible dimensions")
    
    # Initialize result matrix
    C = [[0] * cols_B for _ in range(rows_A)]
    
    for i in range(rows_A):
        for j in range(cols_B):
            for k in range(cols_A):
                C[i][j] += A[i][k] * B[k][j]
    
    return C

def strassen_multiply(A, B):
    """Strassen's algorithm O(n^2.807)"""
    n = len(A)
    
    # Base case
    if n == 1:
        return [[A[0][0] * B[0][0]]]
    
    # Divide matrices into quadrants
    mid = n // 2
    
    A11 = [[A[i][j] for j in range(mid)] for i in range(mid)]
    A12 = [[A[i][j] for j in range(mid, n)] for i in range(mid)]
    A21 = [[A[i][j] for j in range(mid)] for i in range(mid, n)]
    A22 = [[A[i][j] for j in range(mid, n)] for i in range(mid, n)]
    
    B11 = [[B[i][j] for j in range(mid)] for i in range(mid)]
    B12 = [[B[i][j] for j in range(mid, n)] for i in range(mid)]
    B21 = [[B[i][j] for j in range(mid)] for i in range(mid, n)]
    B22 = [[B[i][j] for j in range(mid, n)] for i in range(mid, n)]
    
    # Calculate Strassen's 7 products
    M1 = strassen_multiply(matrix_add(A11, A22), matrix_add(B11, B22))
    M2 = strassen_multiply(matrix_add(A21, A22), B11)
    M3 = strassen_multiply(A11, matrix_subtract(B12, B22))
    M4 = strassen_multiply(A22, matrix_subtract(B21, B11))
    M5 = strassen_multiply(matrix_add(A11, A12), B22)
    M6 = strassen_multiply(matrix_subtract(A21, A11), matrix_add(B11, B12))
    M7 = strassen_multiply(matrix_subtract(A12, A22), matrix_add(B21, B22))
    
    # Calculate result quadrants
    C11 = matrix_subtract(matrix_add(matrix_add(M1, M4), M7), M5)
    C12 = matrix_add(M3, M5)
    C21 = matrix_add(M2, M4)
    C22 = matrix_subtract(matrix_add(matrix_add(M1, M3), M6), M2)
    
    # Combine quadrants
    C = [[0] * n for _ in range(n)]
    for i in range(mid):
        for j in range(mid):
            C[i][j] = C11[i][j]
            C[i][j + mid] = C12[i][j]
            C[i + mid][j] = C21[i][j]
            C[i + mid][j + mid] = C22[i][j]
    
    return C

def matrix_add(A, B):
    """Matrix addition"""
    return [[A[i][j] + B[i][j] for j in range(len(A[0]))] for i in range(len(A))]

def matrix_subtract(A, B):
    """Matrix subtraction"""
    return [[A[i][j] - B[i][j] for j in range(len(A[0]))] for i in range(len(A))]

# Test
A = [[1, 2], [3, 4]]
B = [[5, 6], [7, 8]]
result = matrix_multiply(A, B)
print("Result:", result)  # [[19, 22], [43, 50]]

File: matrix-multiplication/matrix_multiplication.js
function matrixMultiply(A, B) {
    const rowsA = A.length;
    const colsA = A[0].length;
    const rowsB = B.length;
    const colsB = B[0].length;
    
    if (colsA !== rowsB) {
        throw new Error("Cannot multiply matrices: incompatible dimensions");
    }
    
    const C = Array(rowsA).fill(null).map(() => Array(colsB).fill(0));
    
    for (let i = 0; i < rowsA; i++) {
        for (let j = 0; j < colsB; j++) {
            for (let k = 0; k < colsA; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    
    return C;
}

function matrixAdd(A, B) {
    return A.map((row, i) => row.map((val, j) => val + B[i][j]));
}

function matrixSubtract(A, B) {
    return A.map((row, i) => row.map((val, j) => val - B[i][j]));
}

function strassenMultiply(A, B) {
    const n = A.length;
    
    if (n === 1) {
        return [[A[0][0] * B[0][0]]];
    }
    
    const mid = Math.floor(n / 2);
    
    // Divide matrices into quadrants
    const A11 = A.slice(0, mid).map(row => row.slice(0, mid));
    const A12 = A.slice(0, mid).map(row => row.slice(mid));
    const A21 = A.slice(mid).map(row => row.slice(0, mid));
    const A22 = A.slice(mid).map(row => row.slice(mid));
    
    const B11 = B.slice(0, mid).map(row => row.slice(0, mid));
    const B12 = B.slice(0, mid).map(row => row.slice(mid));
    const B21 = B.slice(mid).map(row => row.slice(0, mid));
    const B22 = B.slice(mid).map(row => row.slice(mid));
    
    // Calculate Strassen's 7 products
    const M1 = strassenMultiply(matrixAdd(A11, A22), matrixAdd(B11, B22));
    const M2 = strassenMultiply(matrixAdd(A21, A22), B11);
    const M3 = strassenMultiply(A11, matrixSubtract(B12, B22));
    const M4 = strassenMultiply(A22, matrixSubtract(B21, B11));
    const M5 = strassenMultiply(matrixAdd(A11, A12), B22);
    const M6 = strassenMultiply(matrixSubtract(A21, A11), matrixAdd(B11, B12));
    const M7 = strassenMultiply(matrixSubtract(A12, A22), matrixAdd(B21, B22));
    
    // Calculate result quadrants
    const C11 = matrixSubtract(matrixAdd(matrixAdd(M1, M4), M7), M5);
    const C12 = matrixAdd(M3, M5);
    const C21 = matrixAdd(M2, M4);
    const C22 = matrixSubtract(matrixAdd(matrixAdd(M1, M3), M6), M2);
    
    // Combine quadrants
    const C = Array(n).fill(null).map(() => Array(n).fill(0));
    for (let i = 0; i < mid; i++) {
        for (let j = 0; j < mid; j++) {
            C[i][j] = C11[i][j];
            C[i][j + mid] = C12[i][j];
            C[i + mid][j] = C21[i][j];
            C[i + mid][j + mid] = C22[i][j];
        }
    }
    
    return C;
}

// Test
const A = [[1, 2], [3, 4]];
const B = [[5, 6], [7, 8]];
const result = matrixMultiply(A, B);
console.log("Result:", result); // [[19, 22], [43, 50]]
